### ISSUE TITLE
[Security] You are not allowed to define new elements for path "security.firewalls"

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#22308]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| Symfony version  | -

One of the most frustrating limitations of the Security component config is that you can't define the firewalls in different files:

```
InvalidConfigurationException in PrototypedArrayNode.php line 325:

You are not allowed to define new elements for path "security.firewalls". Please
define all elements for this path in one config file.
```

But this could be so useful! Like in this example:

```yaml
# app/config/dev/security.yml
security:
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false

# app/config/security.yml
security:
    # ...
    firewalls:
        main:
            # ...
```

So, my question for those experts in the internals of the Security component: would be really hard to remove this limitation? Maybe it's a matter of changing a few lines of code but nobody really tried to do that. Thanks!

**Comment by jvasseur** [comment]
AFAIK the reason for this limitation is that the order of firewalls is important since this is the order they are matched. Allowing to split the list in multiple files would make the order harder to predict.

**Comment by stof** [comment]
however, you can perfectly rewrite entirely the configuration of a firewall as soon as it exists in the first loaded file (when the first one providing some firewall config actually)

**Comment by sstok** [comment]
Maybe we should introduce the concept of priorities for a firewall? Or is this to much for a simple use-case.

**Comment by iltar** [comment]
I think one of the reasons this was not possible, was due to the ability to load extra files that might compromise the firewall setup by allowing it to merge with other config.

That being said, it would have to support priorities as @sstok mentioned. Perhaps a debug command would be required as well to view all the compiled information.

**Comment by javiereguiluz** [comment]
I'm closing this because it's not that important and changing it seems difficult and dangerous. Thanks!

**Comment by ro0NL** [comment]
@javiereguiluz can i recall for this issue.

Im not sure how to built dynamic security, right now we end up with multiple configs;
- security_fosuser.yaml
- security_adfs.yaml (external auth)

However now we want a API firewall; meaning 2 additional files;

- security_fosuser_with_api.yaml
- security_adfs_with_api.yaml

I dont like that :) i prefer a priority approach.

**Comment by ro0NL** [comment]
Tried a different approach; alter an existing firewall.

```
You are not allowed to define new elements for path "security.firewalls". Please define all elements for this path in one config file.
```

It's not a _new_ element now =/

**Comment by ro0NL** [comment]
```
Configuration path "security.access_control" cannot be overwritten. You have to define all options for this path, and any of its sub-paths in one configuration section.
```

Come on! This is ridiculous.

**Comment by fabpot** [comment]
That's how Symfony configuration works. We never merge config, just override.

**Comment by ro0NL** [comment]
@fabpot but overriding is also not allowed :thinking: 

AFAIK this is triggered by `disallowNewKeysInSubsequentConfigs()` / `denyOverwrite()` for these specific nodes, solely to overcome the order issue of firewalls.

Basically what i expected if i follow _"all keys have to be defined in the same configuration file"_; id be allowed to overwrite values as it wouldnt affect the order of a firewall.

Basically what im doing with this patch https://github.com/symfony/symfony/issues/24461#issuecomment-334731627 currently.

Either way, i've created an issue for this :)
